부산대 BE_오선재 Step2 - 주문하기

5seonjae opened the pull request:
### 구현 내용
- 주문하기 API 및 서비스 구현
  - 옵션(option) 조회 → 재고 차감 → 주문 저장 → 위시리스트 정리 로직
  - Spring Bean Validation(@Min, @Max) 적용 및 빈 메시지 허용 처리
- Thymeleaf 기반 주문 폼
  - 상품 상세 → /orders/form 이동, 옵션·수량 입력 후 주문
- 주문 내역 리스트(/orders) 및 상세(/orders/{id}) 페이지
  - 페이징(Pageable) 지원, Thymeleaf 테이블 렌더링
- 카카오톡 “나에게 보내기” 메모 API 연동
  - WebClient 설정 및 사용자 정의 템플릿(template_id + template_args) 호출
  - Member → KakaoTokens Embeddable로 토큰 관리, 토큰 만료 시 자동 갱신 지원
- 테스트 보강
  - 단위(Mock) 테스트 추가

---

### 주요 변경점
- Order, Option, Member 엔티티 리팩토링
- OrderRequestDto, OrderViewResponseDto 등 DTO 계층 보강
- Thymeleaf 뷰(order_form.html, order_list.html, order_success.html) 추가
- `.env` 에 `kakao.template-id` 추가
- 예외 처리: InsufficientStockException, ExternalServiceException 매핑 및 글로벌 핸들러

---

### 주문하기 기능 시연 영상

https://www.notion.so/Step2-23e75546dc3080e2a23af3ca2d663eef?source=copy_link

---

### 1단계 피드백 관련 답변

https://github.com/next-step/spring-gift-order/pull/91#discussion_r2232916323

https://github.com/next-step/spring-gift-order/pull/91#discussion_r2232914586

리뷰 부탁드립니다!!!

5seonjae committed:
chore(config): kakao 프로퍼티 키 clientId → client-id, redirectUri → redirect-uri로 통일

- `@Value("${kakao.client-id}")`, `@Value("${kakao.redirect-uri}")` 애노테이션으로 수정
- Thymeleaf 로그인 링크에서 `@environment.getProperty('kakao.client-id')`, `...('kakao.redirect-uri')` 반영

5seonjae committed:
feat(order): Step2 주문하기 기능 명세 및 체크리스트 README.md에 추가

- `POST /api/orders`, `GET /api/orders/{id}`, `GET /api/orders?page=&size=` API 엔드포인트 상세 명세 작성
- 요청 Body 예시 및 응답 예시 JSON (`optionId`, `quantity`, `orderDateTime`, `message` 등) 추가
- Kakao OAuth 인증·인가 시퀀스 다이어그램으로 로그인 → 토큰 발급 → JWT 쿠키 설정 흐름 설명
- 도메인 모델 설계 및 구현을 위한 체크리스트 항목 추가
    - 옵션 재고 검증 및 차감 (`option.decreaseStock()`)
    - 주문 엔티티 저장
    - 주문자 위시리스트 항목 삭제
    - Kakao “나에게 보내기” 메시지 전송
- 예외 처리(404, 409, 502) 및 카카오 API Stub 단위 테스트 TODO 항목 포함

5seonjae committed:
feat(order): 주문 엔티티 및 연관관계 매핑 추가

- `Order` 엔티티 클래스 생성
    - `id`, `quantity`, `message`, `orderDateTime`, `member`, `option` 필드 정의
    - `@Min`, `@Max`, `@Size` 유효성 검증 애노테이션 적용
    - `@CreationTimestamp`로 주문일시 자동 설정
    - 기본 생성자(`protected Order()`) 및 오버로드된 생성자 추가
- `Member`, `Option` 엔티티에 양방향 연관관계 매핑
    - `@OneToMany(mappedBy = "member", cascade = ALL)`를 통한 `Member.orders` 리스트 필드 추가
    - `@OneToMany(mappedBy = "option", cascade = ALL)`를 통한 `Option.orders` 리스트 필드 추가

5seonjae committed:
feat(order): 주문 목록 조회 API 및 페이징 기능 구현

- `OrderController`에 `GET /api/orders` 엔드포인트 추가
- `OrderService`에 `getOrderListForMember` 메서드 구현 및 유효 회원 검증 로직 (`InvalidMemberException`) 적용
- `OrderRepository`에 `findByMemberId(Long, Pageable)` 메서드 정의, `@EntityGraph(attributePaths = {"option", "member"})`로 연관 엔티티 패치 최적화
- `OrderViewResponseDto` 레코드 생성 (주문ID, 상품명, 옵션명, 수량, 메시지, 주문일시)
- `@PageableDefault(size = 5, sort = "id", direction = DESC)`으로 기본 페이징 설정
- `@Transactional(readOnly = true)` 적용으로 조회 전용 트랜잭션 보장

5seonjae committed:
feat(order): 주문 생성 API 및 DTO·서비스 로직 구현

- `POST /api/orders` 엔드포인트 추가 및 `ResponseEntity`로 201 응답 처리
- `OrderRequestDto`에 `@NotNull`, `@Min`, `@Size` 검증 애노테이션 적용
- `OrderResponseDto` 레코드 생성 및 `of()` 팩토리 메서드로 변환 로직 구현
- `OrderService.addOrderForMember` 메서드에서
    - 회원 유효성 검증
    - 옵션 조회 및 재고 차감(`optionService.subtractQuantity`)
    - 주문 엔티티 저장
    - 위시리스트 삭제(`wishRepository.deleteByMemberIdAndProductId`)
- `@Transactional`로 트랜잭션 경계 설정 및 `NoSuchElementException`으로 예외 처리

5seonjae committed:
feat(order): 주문 삭제 API 및 서비스 로직 구현

- `OrderService.deleteOrderForMember` 메서드 추가
    - `validateMember`로 회원 권한 검증
    - `orderRepository.findById`로 주문 존재 여부 확인 후 삭제(`deleteById`)
- `OrderController.deleteOrder` 엔드포인트 추가 (`@DeleteMapping("/{id}")`)
    - `@LoginMember`로 회원 정보 주입
    - 정상 삭제 시 HTTP 204 No Content 응답 처리
- 주문 미존재 시 `NoSuchElementException` 발생으로 404 응답 예상
- 단위 테스트를 위한 TODO: 주문 삭제 성공 및 실패 케이스 검증 코드 작성

5seonjae committed:
feat(order-view): 주문 화면(폼·목록·완료) 기능 구현

- `OrderViewController`에 주문 폼(`/orders/form`), 주문 생성(`/orders` POST → 리다이렉트 `/orders/success`), 주문 목록 조회(`/orders`) 엔드포인트 추가
- Thymeleaf 템플릿 생성
    - `templates/orders/form.html`: 상품 정보 출력, 옵션·수량 선택, 총액 실시간 계산 기능 구현 (JS recalc)
    - `templates/orders/list.html`: 페이징 처리된 주문 내역 테이블, 날짜 포맷팅(`#temporals.format`) 적용
    - `templates/orders/success.html`: 주문 완료 메시지 및 ‘주문 내역 보기’·‘메인으로’ 링크 제공
- `templates/user/detail.html`에 상품 상세에서 주문 폼으로 이동하는 “주문하기” 버튼 추가
- `Product` 엔티티에 옵션 목록 반환용 `getOptions()` 메서드 추가

5seonjae committed:
feat(auth): 카카오 토큰 갱신 및 관리 기능 추가

- `KakaoTokenManager` 구현
    - access token 만료 시 refresh token 으로 재발급
    - refresh token 만료 시 `ReauthorizeRequiredException` 발생 처리
- `KakaoTokenInterceptor` 추가
    - 모든 `/api/**` 요청 전 현재 로그인한 카카오 유저 토큰 만료 검사 및 자동 갱신 보장
- `KakaoTokens` Embeddable 클래스 생성
    - access/refresh 토큰, 만료 시각 관리
    - 컬럼 길이 조정(@AttributeOverride) 적용
- `Member` 엔티티에 `KakaoTokens` 임베딩 및 토큰 조회·갱신 메서드 추가
- `kakaoOAuthService.refreshTokenGrant` 메서드 구현
    - WebClient를 이용한 refresh_token grant flow
- `ReauthorizeRequiredException` 전역 처리기(`@ExceptionHandler`) 추가
    - 토큰 재승인이 필요한 경우 401 UNAUTHORIZED 응답 반환
- `WebConfig` 에 인터셉터 등록 및 `LoginMemberArgumentResolver.resolve` 유틸 메서드 제공

5seonjae committed:
feat(order): 주문 생성 후 카카오 “나에게 보내기” 메시지 전송 및 WishResponseDto 개선

- `KakaoMessageService` 구현
    - `@Value`로 외부 API URL(`kakao.api-url`)·템플릿 ID(`kakao.template-id`) 주입
    - `WebClient` 기반으로 `/v2/api/talk/memo/send` POST 요청 로직 작성
    - `template_args` JSON 직렬화(`ObjectMapper`) 및 폼 데이터 전송 처리
    - 응답 `result_code` 검증 후 실패 시 `ExternalServiceException` 발생
- `ExternalServiceException` 커스텀 예외 클래스 추가
    - 외부 서비스 응답 오류를 `BAD_GATEWAY(502)`로 매핑
- 전역 예외 처리기(`@ExceptionHandler`) 등록
    - `ExternalServiceException` 발생 시 `{"error": "..."} + 502` 응답 반환
- `OrderService.addOrderForMember`에서
    - 주문 저장 직후 `KakaoMessageService.sendOrderMemo(saved, member.getAccessToken())` 호출로 카카오 “나에게 보내기” 알림 전송 로직 반영
- `WishResponseDto.of(WishItem wi)` 정적 팩토리 메서드에 옵션 목록 매핑 로직 추가
    - `wi.getProduct().getOptions().stream().map(option → new OptionResponseDto(...)).toList()` 형태로 옵션 응답 DTO 스트림 처리 적용
- `WishResponseDto` 레코드 내부에 `getOptions()` 게터 메서드 정의 보강
- 관련 단위 테스트 TODO 주석 추가 (주문 알림 전송 성공/실패, WishResponseDto.of 검증)

5seonjae committed:
test(order-service): OrderService 유닛 테스트 추가

- `getOrderListForMember`
    - 주문 없음(0건) 반환 검증
    - 주문 1건 반환 검증
    - 다수 주문 페이징(첫/이전/다음 페이지 플래그) 검증
- `addOrderForMember`
    - 옵션 존재 시 재고 차감, 주문 저장, 위시리스트 삭제, 카카오 메시지 전송 호출 검증
    - 옵션 미존재 시 `NoSuchElementException("옵션을 찾을 수 없습니다.")` 예외 발생 검증
- `deleteOrderForMember`
    - 기존 주문 삭제 시 `deleteById` 호출 검증
    - 주문 미존재 시 `NoSuchElementException("주문을 찾을 수 없습니다.")` 예외 발생 검증
- 공통 회원 검증
    - `null` 멤버 전달 시 `InvalidMemberException("유효하지 않은 회원입니다.")` 예외 발생 검증

5seonjae committed:
test(order): 주문 삭제 테스트에서 불필요한 주석 제거

5seonjae committed:
fix(order): import.sql 이미지 링크 교체 및 주문 폼 총합 계산 로직 보완

- `import.sql` 내 상품 이미지 URL을 기존 프로젝트 내 로컬 경로에서 Imgur 호스팅 링크(`https://i.imgur.com/...`)로 일괄 교체
- `templates/orders/form.html` 주문 폼에 총합 금액 실시간 계산 로직 개선
    - 수량 입력 및 옵션 변경 시 즉시 `단가 × 수량`으로 계산된 금액이 표시되도록 `recalc()` 함수 호출 타이밍 및 초기값 설정 보강

5seonjae committed:
chore(docs): README.md 업데이트 및 체크리스트 완료 표시

5seonjae committed:
fix(option): 옵션 엔티티 수량 최소값 제약조건 1→0으로 완화

- `Option.quantity` 필드의 `@Min` 값을 1에서 0으로 변경하여 옵션 수량을 0개부터 허용
- 검증 메시지 업데이트: `"수량은 0개 이상이어야 합니다."`

5seonjae commented:
지금 테스트가 깨지는 것들이 많이 생겨서 이것들도 빠르게 수정하도록하겠습니다...!

5seonjae committed:
fix(auth): KakaoTokenInterceptor 적용 경로를 `/api/**` → `/api/orders/**`로 제한

- `InterceptorRegistry` 설정에서 전체 API가 아닌 주문 API(`/api/orders/**`)에만 인터셉터 등록
- 불필요한 엔드포인트 호출 시 토큰 검사 및 갱신 로직 실행 방지

Gyuchool commented on README.md:
```diff
@@ -137,4 +140,74 @@ https://kauth.kakao.com/oauth/authorize
+
+## 📦 Step2 - 주문하기
+
+### 🧩 기능 명세
```
정리를 깔끔하게 하셨군요👍

Gyuchool commented on src/main/java/gift/controller/view/OrderViewController.java:
```diff
@@ -0,0 +1,94 @@
+    ) {
+        Product product = productService.getProductById(productId);
+        model.addAttribute("product", product);
+        model.addAttribute("orderReq", new OrderRequestDto(null, 1, ""));
```
null, 1, "" 모두 어떤 것을 의미하는지 모르겠네요🙄 매개변수를 받아서 처리해야하는 걸까요? 의미가 분명하면 좋을 것 같습니다!

5seonjae replied:
피드백 감사드립니다! 😊

이전처럼 의미가 불명확한 `new OrderRequestDto(null, 1, "")` 대신, 아래와 같이 동적으로 기본 옵션 ID 를 구해서 DTO 팩토리 메서드에 넘기도록 리팩토링했습니다.

```java
@GetMapping("/orders/form")
    public String orderForm(
        @RequestParam Long productId,
        Model model,
        RedirectAttributes redirectAttributes
    ) {
        Product product = productService.getProductById(productId);
        if (product.getOptions().isEmpty()) {
            redirectAttributes.addFlashAttribute(
                "errorMessage",
                "상품에 옵션이 없어서 주문을 할 수 없습니다. 옵션이 추가되면 주문해주세요."
            );
            return "redirect:/products/" + productId;
        }

        Long defaultOptionId = product.getOptions().get(0).getId();

        model.addAttribute("product", product);
        model.addAttribute(
            "orderReq",
            OrderRequestDto.defaultOrderRequestDto(defaultOptionId)
        );
        return "orders/form";
    }
```

```java
public static OrderRequestDto defaultOrderRequestDto(Long defaultOptionId) {
        return new OrderRequestDto(defaultOptionId, 1, "");
    }
```

```java
<div th:if="${errorMessage}" class="alert alert-danger" th:text="${errorMessage}"></div>
```

이제 `defaultOrderRequestDto` 메서드만 보면 “첫 번째 옵션 ID, 기본 수량 ( 1 ), 기본 메모 ( 빈 문자열 ) 를 담은 DTO” 라는 의도가 즉시 와 닿으므로, null, 1, "" 리터럴이 갖고 있던 모호함을 완전히 해소할 수 있을 것 같습니다.

추가로 상품 옵션이 없을 경우에는 상품 상세 페이지에서 주문하기 버튼을 누를 시에 "상품에 옵션이 없어서 주문을 할 수 없습니다. 옵션이 추가되면 주문해주세요." 메세지를 띄우고 다시 상품 상세 페이지로 리다이렉트 하는 방식으로 편의성을 개선했습니다.

<img width="1032" height="842" alt="image" src="https://github.com/user-attachments/assets/f60e3ece-cba3-4ab5-929f-68fbcf802a06" />

감사합니다! 😊

Gyuchool commented on src/main/java/gift/dto/api/OrderRequestDto.java:
```diff
@@ -0,0 +1,36 @@
+public class OrderRequestDto {
+
+    @NotNull(message = "옵션 ID는 필수입니다.")
+    Long optionId;
```
접근제어자를 안쓰신 이유가 있을까요?🙄

5seonjae replied:
리뷰 감사드립니다!

접근제어자를 지정하지 않은 건 깜빡한 실수였습니다. 
DTO 필드는 외부에서 직접 변경되지 않도록 `private` 으로 감싸는 게 맞으니, 아래와 같이 수정하여 곧 푸시하겠습니다:

```java
public class OrderRequestDto {

    @NotNull(message = "옵션 ID는 필수입니다.")
    private Long optionId;

    @NotNull(message = "옵션 수량은 필수입니다.")
    @Min(value = 1, message = "한 개 이상 담을 수 있습니다.")
    private Integer quantity;

    @Size(max = 200, message = "요청 메시지는 200자 이하만 가능합니다.")
    private String message;

    // …생성자·게터는 그대로 유지
}
```

Gyuchool commented on src/main/java/gift/dto/api/OrderRequestDto.java:
```diff
@@ -0,0 +1,36 @@
+    @Min(value = 1, message = "한 개 이상 담을 수 있습니다.")
+    Integer quantity;
+
+    @Size(max = 200,message = "요청 메시지는 200자 이하만 가능합니다.")
```
validation 설정👍

Gyuchool commented on src/main/java/gift/dto/api/OrderRequestDto.java:
```diff
@@ -0,0 +1,36 @@
+
+    @NotNull(message = "옵션 수량은 필수입니다.")
+    @Min(value = 1, message = "한 개 이상 담을 수 있습니다.")
+    Integer quantity;
```
int로 설정한것과 Integer로 설정한 것과 무슨 차이가 있을까요?🙄

5seonjae replied:
안녕하세요 리뷰어님!

피드백 주신 int vs Integer 차이에 대해 답변드리겠습니다.

1. int(primitive) 의 특성

- 기본값 0

  - 스프링이 바인딩할 때 요청값이 없으면 `quantity` 는 0 으로 세팅됩니다.

- null 불가

  - 자바 원시 타입이라 절대 null이 될 수 없으므로, `@NotNull` 검증이 아예 의미가 없습니다.

2. Integer(wrapper) 를 쓴 이유

- null 허용

  - 클라이언트가 `quantity` 필드를 아예 보내지 않았을 때 `null` 로 바인딩되어, “값 누락” 상태를 감지할 수 있습니다.

- Bean Validation

  - `@NotNull` 어노테이션이 정상 동작해서 “옵션 수량은 필수입니다”라는 메시지로 400 응답을 줄 수 있습니다.

Integer 대신 primitive 를 쓰면서 기본값 ( 예 : 1 ) 으로 처리하더라도 무방하지만, 그럴 경우 “아예 누락된 요청”과 “0 으로 보내진 요청”을 구분할 수 없으므로 엄격한 API 검증을 위해 wrapper 타입+검증으로 구현했습니다.

Gyuchool replied:
👍 auto boxing에 대해서도 학습하면 좋겠습니다!

Gyuchool commented on src/main/java/gift/dto/api/OrderResponseDto.java:
```diff
@@ -0,0 +1,43 @@
+        return message;
+    }
+
+    public static OrderResponseDto of(Order order) {
```
선재님은 언제 정적팩토리 메서드 네이밍으로 of를 사용하실까요?

5seonjae replied:
좋은 지적 감사합니다!

솔직히 그동안 정적 팩토리 메서드 네이밍에 크게 신경 쓰지 않았는데, 리뷰어님의 피드백을 보고 정적 팩토리 메서드의 네이밍에 대해 다시 생각을 해보았습니다.

- `of(...)` : 같은 타입끼리 포장 ( wrapping ) 할 때

- `from(...)` : 다른 타입 간 변환 ( conversion ) 을 명시할 때

라는 관례를 적용해 보니, `OrderViewResponseDto` 는 Order 엔티티를 DTO 로 변환하는 용도이므로 `of` 보다는 `from(Order order)` 이 더 어울린다고 판단했습니다.

그래서 다음과 같이 수정하도록 하겠습니다.

```java
public static OrderViewResponseDto from(Order order) {
        return new OrderViewResponseDto(
            /* 매핑 로직 */
        );
    }
```

다음부터는 네이밍에 더 신경써서 구현하도록 하겠습니다!
혹시 더 구체적인 네이밍 제안이나, 다른 컨벤션이 있으시면 알려주세요! 😊

Gyuchool replied:
보통 이팩티브 자바에서 나온 관례를 따르긴 합니다 :) 선재님만의 네이밍 규칙이 있으신지 궁금해서 여쭤봤습니다..!

https://kang-james.tistory.com/entry/Effective-Java-%EC%83%9D%EC%84%B1%EC%9E%90-%EB%8C%80%EC%8B%A0-Static-Factory-Method%EB%A5%BC-%EA%B3%A0%EB%A0%A4%ED%95%98%EB%9D%BC#google_vignette

Gyuchool commented on src/main/java/gift/entity/KakaoTokens.java:
```diff
@@ -0,0 +1,70 @@
+        return new KakaoTokens(
+            kakaoTokenResponseDto.accessToken(),
+            kakaoTokenResponseDto.refreshToken(),
+            LocalDateTime.now().plusSeconds(kakaoTokenResponseDto.expiresIn()),
+            LocalDateTime.now().plusDays(30)
```
30일의 의미는 무엇일까요?

5seonjae replied:
리뷰 감사합니다!

`plusDays(30)` 으로 리프레시 토큰 만료 시점을 잡은 이유는,
저번 라이브 강의에서 공유된 카카오 OAuth 전략을 참고했기 때문입니다.

<img width="953" height="244" alt="image" src="https://github.com/user-attachments/assets/7a05360a-97b9-4ef6-ab07-9fdbbc359b74" />

>응답은 [토큰 요청](https://developers.kakao.com/docs/latest/ko/kakaologin/rest-api#request-token)과 마찬가지로 JSON 객체로 전달됩니다. 응답 중 refresh_token 값은 요청 시 사용된 리프레시 토큰의 만료 시간이 1개월 미만으로 남았을 때만 갱신되어 전달됩니다. 따라서 refresh_token과 refresh_token_expires_in은 결과 값에 포함되지 않을 수 있다는 점을 응답 처리 시 유의해야 합니다.

Kakao 공식 문서 / 강의 자료 에 따르면 리프레시 토큰의 기본 만료 기간은 최소 30일이기 때문에 토큰 저장 시점으로부터 30일 뒤를 만료 시한으로 설정하는 것이 맞다고 판단했습니다.

하지만, 다시 생각해보니 30일로 하드코딩 하는 것보다 리프레시 토큰 만료 시간 ( `refresh_token_expires_in` ) 을 받아오는 것이 더 맞다고 생각했습니다. 그래서 다음과 같이 수정하려 합니다.

```java
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public record KakaoTokenResponseDto(
    String accessToken,
    String refreshToken,
    long expiresIn,
    long refreshTokenExpiresIn,
    String tokenType
) {}
```

```java
public static KakaoTokens from(KakaoTokenResponseDto kakaoTokenResponseDto) {
        return new KakaoTokens(
            kakaoTokenResponseDto.accessToken(),
            kakaoTokenResponseDto.refreshToken(),
            LocalDateTime.now().plusSeconds(kakaoTokenResponseDto.expiresIn()),
            LocalDateTime.now().plusDays(kakaoTokenResponseDto.refreshTokenExpiresIn())
        );
    }
```

이렇게 하면 하드코딩된 30 일이 아니라, 카카오가 내려주는 정확한 만료 기간을 그대로 반영할 수 있습니다. 
추가 의견 있으시면 말씀 부탁드립니다! 😊

Gyuchool commented on src/main/java/gift/entity/Member.java:
```diff
@@ -32,9 +36,19 @@ public class Member {
 
+    @Embedded
+    @AttributeOverrides({
+        @AttributeOverride(name = "accessToken", column = @Column(length = 2048)),
```
2048로 길이를 정하신 이유가 있으실까요?

5seonjae replied:
accessToken 필드에 `length = 2048`을 준 이유는, OAuth2/JWT 액세스 토큰이 Base64 인코딩된 문자열 형태로 내려오기 때문에 길이가 꽤 길어질 수 있기 때문입니다.

예컨대 헤더·페이로드·서명 세 부분을 합친 JWT 토큰은 보통 300~800자 정도 되지만, 앞뒤로 확장되거나 추가 클레임이 붙으면 1,000자 이상 넘어갈 수도 있습니다.

그래서 안정 마진을 두고 2,048자로 넉넉하게 잡아두었습니다.

이렇게 하면 토큰 길이 초과로 인한 저장 오류를 방지할 수 있을 것 같습니다! 
추가 의견 있으시면 말씀 부탁드립니다! 😊

Gyuchool commented on src/main/java/gift/entity/Member.java:
```diff
@@ -32,9 +36,19 @@ public class Member {
     @OneToMany(mappedBy = "member", cascade = CascadeType.ALL, orphanRemoval = true)
     private List<WishItem> wishItems = new ArrayList<>();
 
+    @OneToMany(mappedBy = "member", cascade = CascadeType.ALL)
```
여기에는 orphanRemoval를 설정 안하신 이유가 있으실까요?

5seonjae replied:
리뷰 감사드립니다! 😊

cascade 와 orphanRemoval 설정 관련해서 아래 글을 참고했습니다.

https://tecoble.techcourse.co.kr/post/2021-08-15-jpa-cascadetype-remove-vs-orphanremoval-true/

`orders` 컬렉션에는 `orphanRemoval = true` 를 걸지 않은 데에는 두 가지 이유가 있습니다

### 1. Order는 WishItem과 달리 독립된 Aggregate

Order는 주문 이력을 관리하는 핵심 도메인으로, Member 뿐 아니라 옵션 ( Option ) 엔티티와도 연결되어 있어, “컬렉션에서 빠진다”는 이유만으로 곧바로 DB에서 삭제되는 건 적절치 않다고 생각했습니다.

### 2. `orphanRemoval` 의 부작용 방지

`orphanRemoval=true` 는 `member.getOrders().remove(someOrder);` 처럼 컬렉션에서 분리되는 순간, 해당 Order 를 **고아 ( Orphan )** 로 보고 즉시 삭제합니다.

실제 도메인에서는 “조회용으로 List 에서 제외만 하고, DB 에는 남겨두고 싶다”는 경우가 많기 때문에, 일시적으로 컬렉션 조작만으로 주문 자체가 사라지면 안 됩니다.

---

대신, Member를 삭제할 때만 CascadeType.REMOVE 로 연쇄 삭제되도록 두어, Member가 탈퇴되면 그에 속한 모든 주문도 함께 지워지고 단순히 컬렉션에서 빼는 작업에는 영향을 받지 않도록 설계했습니다.

혹시 더 논의할 부분이 있으면 편하게 말씀 주세요! 👍
이렇게 구현하는 것이 맞다면 wishItem 엔티티에도 동일하게 구현하겠습니다.

Gyuchool commented on src/main/java/gift/entity/Order.java:
```diff
@@ -0,0 +1,94 @@
+
+    protected Order() {}
+
+    public Order(
```
생성자 체이닝을 적용해볼까요?

5seonjae replied:
리뷰어님, 꼼꼼한 지적 감사합니다!

말씀하신 대로 `orderDateTime` 을 `@CreationTimestamp` 에 맡기고, 생성자 중복을 제거하기 위해 생성자 체이닝을 적용해 보았습니다.

```java
@Entity
@Table(name = "orders")
public class Order {
    // … 생략 …

    @Column(nullable = false, updatable = false)
    @CreationTimestamp
    private LocalDateTime orderDateTime;

    protected Order() {}

    //  메인 생성자: 모든 필드를 주입(테스트·데이터 마이그레이션용)
    public Order(
        int quantity,
        String message,
        LocalDateTime orderDateTime,
        Member member,
        Option option
    ) {
        this.quantity      = quantity;
        this.message       = message;
        this.orderDateTime = orderDateTime;
        this.member        = member;
        this.option        = option;
    }

    // 보조 생성자: orderDateTime 은 Hibernate 가 자동 설정하도록 생략
    public Order(int quantity, String message, Member member, Option option) {
        this(quantity, message, LocalDateTime.now(), member, option);
        //  orderDateTime 필드는 @CreationTimestamp가 덮어씁니다
    }
}
```

`@CreationTimestamp` 덕분에 Hibernate 가 INSERT 시점에 자동으로 현재 시각을 채워 주므로, 생성자에서 직접 `now()` 를 호출해 주입해도 DB 에는 삽입 직전 자동값이 적용됩니다.

메인 생성자 ( 5-arg ) 는 테스트나 데이터 마이그레이션 시 ‘임의의 시간’을 주입할 수 있도록 열어 두었고, 보조 생성자 ( 4-arg ) 에서 메인 생성자를 호출해 중복 필드 할당을 방지했습니다.

이렇게 구현하면 중복 코드 없이, `orderDateTime` 은 Hibernate 에 완전히 위임하면서도 생성자 체이닝의 장점을 살릴 수 있습니다.
추가 의견 있으시면 언제든 말씀 부탁드립니다! 😊

Gyuchool commented on src/main/java/gift/entity/Product.java:
```diff
@@ -79,4 +79,8 @@ public int getPrice() {
     }
+
+    public List<Option> getOptions() {
+        return options;
```
방어적 복사를 적용하면 어떤 장점이 있을까요?

참고 자료: https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/

5seonjae replied:
좋은 자료 공유 감사합니다!

방어적 복사를 적용할 때 얻을 수 있는 장점은 크게 다음과 같습니다.

1. 엔티티 캡슐화 유지

내부 options 리스트의 참조를 외부에 직접 노출하지 않음으로써, 외부 코드가 product.getOptions().add(...)처럼 실수로 엔티티 상태를 변경하는 것을 막을 수 있습니다.

2. 상태 일관성 보장

방어적 복사로 반환된 컬렉션은 원본과 분리된 객체이기 때문에, 외부에서 리스트를 변경하더라도 내부 options에는 전혀 영향을 주지 않습니다.

3. 명시적 불변(또는 안전한 뷰) 제공

new ArrayList<>(options) 처럼 복사본을 반환하거나, Collections.unmodifiableList(options) 를 쓰면 “여기서는 수정할 수 없습니다”라는 의도가 가장 분명히 드러납니다.

4. 스레드 안전성 향상

멀티스레드 환경에서 한 스레드가 반환받은 리스트를 수정해도, 다른 스레드가 보고 있는 내부 리스트는 그대로 유지되어 경합(race)이나 예측 불가능한 버그를 예방할 수 있습니다.

5. 디버깅·유지보수 편의성

“어디서 리스트가 바뀌었지?” 하고 추적하던 중, 외부 변경이 불가능함을 보장하면 문제 범위를 엔티티 내부로 좁혀 디버깅이 쉬워집니다.

```java
// 방어적 복사 예시 1: 복사본 반환
public List<Option> getOptions() {
    return new ArrayList<>(options);
}

// 방어적 복사 예시 2: 수정 불가 뷰 반환
public List<Option> getOptions() {
    return Collections.unmodifiableList(options);
}
```

위 두 방법 모두 외부에서 내부 리스트를 건드리지 못하게 막아 주므로, 엔티티의 불변성과 상태 일관성을 지키는 데 큰 도움이 됩니다.

그래서 다음과 같이 수정하려고 합니다.

```java
public List<Option> getOptions() {
    return new ArrayList<>(options);
}
```

추가 의견 있으시면 언제든 말씀 부탁드립니다! 😊

Gyuchool replied:
정리를 너무 잘해주셨네요💯

Gyuchool commented on src/main/java/gift/interceptor/KakaoTokenInterceptor.java:
```diff
@@ -0,0 +1,38 @@
+        HttpServletResponse res,
+        Object handler
+    ) {
+        Member member = resolver.resolve(req);
```
인터셉터의 역할은 무엇일까요? resolver와 인터셉터는 어떤 차이가 있나요?

5seonjae replied:
인터셉터 ( Interceptor ) 와 리졸버 ( HandlerMethodArgumentResolver ) 는 모두 스프링 MVC의 확장 포인트이지만, 역할과 적용 시점이 다릅니다.

### 인터셉터 ( HandlerInterceptor )

1. 동작 지점

- 클라이언트 요청이 디스패처 서블릿에 도달한 직후부터, 컨트롤러 호출 전·후·완료 시점까지 훑어갑니다.

2.  주요 책임

- 인증·인가, 로깅, 성능 측정, 공통 헤더 설정 등 요청 단위의 전반적인 처리
- `preHandle()` → `postHandle()` → `afterCompletion()` 메서드로 단계별 제어

3. 특징

- URL 패턴 ( `/api/**` ) 단위로 적용 가능
- 컨트롤러 메서드나 파라미터 구조를 몰라도 동작

### 아규먼트 리졸버 ( HandlerMethodArgumentResolver )


1. 동작 지점

- 컨트롤러의 핸들러 메서드를 호출하기 직전, 메서드 시그니처에 선언된 파라미터 하나하나를 바인딩할 때

2.  주요 책임

- `@LoginMember Member member` 같은 커스텀 애노테이션이나 특정 타입 파라미터를 어떻게 생성 ( 또는 주입 ) 할지 결정

3. 특징

- 메서드 인자 단위로 동작하며, 리졸버가 `supportsParameter()` 를 반환하면 `resolveArgument()` 를 통해 값을 공급
- 호출 시점에는 이미 `preHandle()` 같은 인터셉터는 끝난 상태이므로, 인터셉터에서 설정해 둔 `request attribute` 를 꺼내서 쓰기 좋습니다.

### 정리

- 인터셉터 : HTTP 요청 전체 라이프사이클을 훑으며, 토큰 검증 등 글로벌·공통 처리

- 리졸버 : 컨트롤러 메서드 호출 직전에 특정 파라미터에 들어갈 값을 바인딩

이 프로젝트에서 `KakaoTokenInterceptor` 는 먼저 요청에서 토큰을 꺼내 검증하고, `LoginMemberArgumentResolver` 는 그 결과를 `@LoginMember` 파라미터로 주입하는 역할 분담을 하고 있습니다.
이렇게 책임을 분리하면 인증 로직은 인터셉터에, 파라미터 바인딩 로직은 리졸버에 모아 코드의 가독성·유지보수성을 높일 수 있습니다.

추가 질문이 있으시면 언제든 알려주세요! 감사합니다. 😊

Gyuchool replied:
> 호출 시점에는 이미 preHandle() 같은 인터셉터는 끝난 상태이므로, 인터셉터에서 설정해 둔 request attribute 를 꺼내서 쓰기 좋습니다.

설명해주신대로 인터셉터에서 리졸버를 호출할 일이 없어야 할것 같아요! 두개의 역할이 명확히 구별이 안되어보이네요🤔

5seonjae replied:
꼼꼼한 지적 감사합니다!
피드백 주신 대로 Interceptor가 Resolver를 직접 호출해서는 안 되고, 두 역할이 분리되어야 한다고 판단하여 아래처럼 리팩토링했습니다.

```java
@Component
public class KakaoTokenInterceptor implements HandlerInterceptor {

    private final KakaoTokenManager tokenManager;
    private final MemberExtractor memberExtractor;

    public KakaoTokenInterceptor(
        KakaoTokenManager tokenManager,
        MemberExtractor memberExtractor
    ) {
        this.tokenManager = tokenManager;
        this.memberExtractor = memberExtractor;
    }

    @Override
    public boolean preHandle(
        HttpServletRequest req,
        HttpServletResponse res,
        Object handler
    ) {
        if (!(handler instanceof HandlerMethod hm) ||
            Arrays.stream(hm.getMethodParameters())
                .noneMatch(p -> p.hasParameterAnnotation(LoginMember.class))) {
            return true;
        }

        Member member = memberExtractor.extractMember(req);
        if (member.isKakaoUser()) {
            tokenManager.ensureAccessToken(member.getId());
        }
        return true;
    }
}
```

```java
@Component
public class LoginMemberArgumentResolver
    implements HandlerMethodArgumentResolver {

    private final MemberExtractor memberExtractor;

    public LoginMemberArgumentResolver(
        MemberExtractor memberExtractor
    ) {
        this.memberExtractor = memberExtractor;
    }

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.hasParameterAnnotation(LoginMember.class)
                && parameter.getParameterType().equals(Member.class);
    }

    @Override
    public Object resolveArgument(
        MethodParameter parameter,
        ModelAndViewContainer mavContainer,
        NativeWebRequest webRequest,
        WebDataBinderFactory binderFactory
    ) {
        HttpServletRequest req = webRequest.getNativeRequest(HttpServletRequest.class);
        return memberExtractor.extractMember(req);
    }
}
```

```java
@Service
public class MemberExtractor {

    private final TokenService tokenService;
    private final MemberRepository memberRepository;
    private final BearerAuthHeaderParser bearerAuthHeaderParser;
    private final TokenExtractor tokenExtractor;

    public MemberExtractor(
        TokenService tokenService,
        MemberRepository memberRepository,
        BearerAuthHeaderParser bearerAuthHeaderParser,
        TokenExtractor tokenExtractor
    ) {
        this.tokenService = tokenService;
        this.memberRepository = memberRepository;
        this.bearerAuthHeaderParser = bearerAuthHeaderParser;
        this.tokenExtractor = tokenExtractor;
    }

    public Member extractMember(HttpServletRequest req) {
        String header = tokenExtractor.extractBearerHeader(req);
        String token = bearerAuthHeaderParser.extractBearerToken(header);
        Claims claims = tokenService .parseClaims(token);
        Long memberId = Long.valueOf(claims.getSubject());
        return memberRepository.findById(memberId)
            .orElseThrow(() -> new IllegalStateException("존재하지 않는 회원입니다."));
    }
}
```

이렇게 하면 Interceptor 와 Resolver 가 서로를 직접 호출하지 않고, 각자가 맡은 책임을 분명히 지킬 수 있다고 생각합니다.
그리고, Interceptor 와 Resolver 에 필요한 extractMember 메서드를 MemberExtractor 클래스에 따로 넣어서 중복 코드를 완전히 제거했습니다.
감사합니다! 😊

Gyuchool replied:
토큰에 카카오 유저 여부를 나타내는 역할(Role)을 포함시켜 전달하면, 인터셉터에서는 해당 Role만으로 처리 여부를 판단할 수 있겠네요! 이렇게 하면 인터셉터에서 별도로 DB 접근을 하지 않아도 되고, 필요한 DB 조회는 이후 단계인 Argument Resolver에서 수행하게 되어 책임이 분리될것 같습니다:)

Gyuchool commented on src/main/java/gift/repository/OrderRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface OrderRepository extends JpaRepository<Order, Long> {
+
+    @EntityGraph(attributePaths = {"option", "member"})
```
entityGraph를 사용하신 이유가 있을까요?

5seonjae replied:
이전에 제가 이 부분과 관련된 피드백을 받았던 적이 있었습니다.

https://github.com/next-step/spring-gift-enhancement/pull/197#discussion_r2215296872

`@EntityGraph(attributePaths = {"option", "member"})` 를 달아둔 이유는, Order 엔티티의 `option` 과 `member` 연관관계를 기본 LAZY 로딩으로 두었을 때 `findAll()` 등으로 여러 주문을 조회하면서 각각 `order.getOption()`, `order.getMember()` 를 호출하면 SQL N+1 문제가 발생하기 때문입니다.

`@EntityGraph` 를 걸어두면 JPA가 한 번의 JOIN 조회로 `orders` + `option`, `member` 데이터를 모두 가져와 불필요한 추가 쿼리를 막고 응답 속도를 크게 개선할 수 있습니다.

따라서 N+1 방지와 성능 최적화를 위해 `@EntityGraph` 를 적용하였습니다.
추가 의견 있으시면 언제든 말씀 부탁드립니다! 😊

Gyuchool replied:
EntityGraph외에도 fetch join, 혹은 fetch Type을 EAGER로 거는 방식이 있을텐데 entityGraph를 선택하신 이유가 궁금했습니다!

5seonjae replied:
질문 감사합니다!

`@EntityGraph` 를 선택한 이유는 다음과 같습니다

1. 도메인 설계 건드리지 않기

- `FetchType.EAGER` 로 바꾸면 전역적으로 항상 option·member 를 함께 로딩하게 되어, 다른 조회 시에도 불필요한 조인 부하가 발생할 수 있습니다.
- 반면 `@EntityGraph` 는 해당 리포지토리 메서드 한정으로만 즉시 로딩을 적용해 주므로, 전체 애플리케이션 성능에 미치는 영향을 최소화할 수 있습니다.

2. JPQL `fetch join` 보다 선언적 재사용성이 좋다.

- 직접 JPQL 로 `JOIN FETCH` 를 쓰면 그 쿼리 한 가지 경우에만 최적화가 적용됩니다.
- `@EntityGraph` 는 인터페이스 메서드 위에 붙여 두기만 하면, 메서드 이름과 파라미터만 달라질 뿐 어디서든 동일한 최적화 전략을 재사용할 수 있습니다.

3. 비즈니스 로직과 쿼리 분리

- 서비스 레이어에서 “N+1 방지용 fetch join 쿼리”를 직접 호출하지 않아도 되고, 레포지토리 인터페이스 선언만으로 충분하기 때문에, 코드 가독성과 유지보수성이 높아집니다.

4. LAZY 기본 유지로 예측 가능한 성능 제어

- 도메인 관계는 기본 LAZY 를 그대로 두어, 정말 필요한 시점에만 조인하도록 제어하고, 특정 API ( 여기서는 주문 목록 조회 ) 에서만 `@EntityGraph` 로 최적화하는 방식이 가장 유연하고 안전하다고 판단했습니다.

이렇게 하면 “LAZY 기본 + 한정적 즉시 로딩”의 장점과 “쿼리 재사용성”을 모두 확보할 수 있습니다.
추가 의견 있으시면 언제든 말씀 부탁드립니다! 😊

Gyuchool commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -0,0 +1,85 @@
+            form.add("template_id", String.valueOf(templateId));
+            form.add("template_args", buildTemplateArgs(order));
+
+            Map<?, ?> resp = kakaoClient.post()
```
`?` 와일드카드 사용을 지양해야하는 이유가 무엇일까요?

5seonjae replied:
와일드카드 ( `Map<?, ?>` ) 사용을 지양해야 하는 이유는 다음과 같습니다.

1. 타입 안전성(type safety) 확보

2. 가독성·유지보수성 저하

저번에 제네릭 타입 추론 관련 피드백 주셨는데, 이번에 구현하면서 이를 놓쳤던 것 같습니다...
그래서 `Map<?,?>` 로 받던 부분을 구체적인 제네릭( `Map<String, Object>` ) 으로 바꿨습니다.

```java
// 기존
Map<?, ?> resp = kakaoClient.post()…

// 변경
Map<String, Object> resp = kakaoClient.post()
```

와일드카드를 제거해 키는 `String`, 값은 `Object` 라는 타입이 명확해졌습니다.

다시 한 번 꼼꼼한 피드백 감사드립니다! 😊

5seonjae committed:
refactor(auth): ArgumentResolver·Interceptor 로직 개선 및 적용 범위 확장, WishControllerTest 보강

- `LoginMemberArgumentResolver.resolveArgument` 리팩토링: 공통 로직을 `extractMember()`로 분리하여 가독성·재사용성 향상
- `KakaoTokenInterceptor.preHandle` 수정: `handler instanceof HandlerMethod` 검사 및 `@LoginMember` 파라미터 유무 확인 후에만 토큰 갱신 로직 실행
- `WebConfig.addInterceptors` 인터셉터 적용 경로에 `/api/wishes/**`, `/api/products/**` 추가하여 주문뿐 아니라 위시·상품 API에도 토큰 검사 적용
- `WishControllerTest.setup`에 `loginMemberArgumentResolver.resolve(any())` 스텁 추가:
    - `Authorization` 헤더 미존재 시 `MissingAuthorizationHeaderException`
    - 형식 불일치 시 `InvalidAuthorizationHeaderException` 발생 검증 로직 보강

5seonjae commented:
테스트 수정 완료했습니다!

추가로 인터셉터와 리졸버 기능도 수정했습니다

Gyuchool reviewed (changes requested):
안녕하세요 선재님!
빠르게 구현해주셨군요👍 시연 영상까지 보여주셔서 플로우 이해하기 너무 쉬웠습니다!
다만, 제가 직접 했을때는 주문했을때 403에러가 나오네요..?🥲 확인 한 번 부탁드리겠습니다!

<img width="1311" height="145" alt="스크린샷 2025-07-29 오전 12 04 37" src="https://github.com/user-attachments/assets/c40c6c73-bafb-437f-8275-dd2af0b76cf2" />

몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 다시 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
끝까지 화이팅입니다💪

5seonjae committed:
feat(auth): 카카오 로그인 OAuth 스코프에 talk_message 권한 추가

- 로그인 링크의 `scope` 파라미터에 `talk_message`를 포함시켜 사용자 프로필 닉네임 조회(`profile_nickname`)와 함께 메시지 전송 권한 요청 가능하도록 수정

5seonjae committed:
refactor(order-view): 주문 폼 OrderRequestDto 기본 생성자 사용 및 기본값 설정 추가

- `OrderViewController#orderForm`에서 `new OrderRequestDto(null, 1, "")` → `new OrderRequestDto()`로 변경
- `OrderRequestDto`에 no-args 생성자 추가하여 `quantity = 1`, `message = ""` 기본값 설정

5seonjae committed:
fix(dto): OptionRequestDto, OptionSubtractRequestDto, OrderRequestDto 필드에 private 접근 제어자 추가

- 세 가지 DTO 클래스의 필드를 모두 `private`으로 선언하여 캡슐화 강화
- 외부에서 필드에 직접 접근 불가하도록 접근 제어자 일관성 유지

5seonjae committed:
refactor(dto): 정적 팩토리 메서드 이름 of→from으로 변경

- `OptionResponseDto.of(...)` → `OptionResponseDto.from(...)`으로 메서드명 변경
- `OrderResponseDto.of(...)` → `OrderResponseDto.from(...)`으로 메서드명 변경
- `WishResponseDto.of(...)` → `WishResponseDto.from(...)`으로 메서드명 변경
- `OrderViewResponseDto.of(...)` → `OrderViewResponseDto.from(...)`으로 메서드명 변경
- 관련 호출부 및 단위 테스트 코드도 함께 수정하여 메서드명 일관성 유지

5seonjae committed:
fix(auth): KakaoTokens.from 메서드에서 refreshTokenExpiresIn 동적 반영 및 KakaoTokens 정적 팩토리 메서드 이름 `of` → `from` 변경

- `KakaoTokens.from(KakaoTokenResponseDto)` 로직에서 하드코딩된 `LocalDateTime.now().plusDays(30)` → `LocalDateTime.now().plusDays(kakaoTokenResponseDto.refreshTokenExpiresIn())` 로 변경
- `KakaoTokenResponseDto`에 추가된 `refreshTokenExpiresIn` 필드를 토큰 만료일 계산에 반영
- `KakaoTokens` 정적 팩토리 메서드 이름 `of` → `from` 변경

5seonjae committed:
refactor(order): Order 생성자 체이닝 적용하여 주문 일시 자동 설정

- `Order(int quantity, String message, Member member, Option option)` 생성자에서 직접 필드 할당 방식 → `this(quantity, message, LocalDateTime.now(), member, option)` 호출로 변경
- 주문 생성 시 `orderDateTime`이 자동으로 현재 시각으로 설정되도록 개선

5seonjae committed:
refactor(product): getOptions()에 방어적 복사 적용

- `getOptions()` 메서드에서 내부 `options` 리스트를 직접 반환하던 방식 → `new ArrayList<>(options)`로 복사본 반환하도록 변경하여 캡슐화 및 불변성 강화

5seonjae committed:
fix(message): sendOrderMemo 응답 타입 Map<?,?> → Map<String, Object>로 구체화

- `bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {})`로 제네릭 타입을 명시하여 응답 맵 처리 안정성 강화

5seonjae committed:
fix(order-view): 주문 폼 옵션 유무 검증 및 기본 옵션 자동 설정

- `OrderViewController#orderForm`에 `RedirectAttributes` 파라미터 추가
- 상품에 옵션이 없을 경우 플래시 속성 `errorMessage` 전달 후 `/products/{productId}`로 리다이렉트
- 첫 번째 옵션 ID를 이용해 `OrderRequestDto.defaultOrderRequestDto(defaultOptionId)`로 주문 요청 DTO 기본값 설정
- `OrderRequestDto`에 `defaultOrderRequestDto(Long)` 정적 팩토리 메서드 추가
- `templates/products/user/detail.html`에 `errorMessage`를 출력하는 alert 영역 추가

Gyuchool commented on src/main/java/gift/controller/view/OrderViewController.java:
```diff
@@ -0,0 +1,108 @@
+            return "redirect:/products/" + productId;
+        }
+
+        Long defaultOptionId = product.getOptions().get(0).getId();
```
해당 시점에 쿼리가 어떻게 발생하나요? 
한 번만 쿼리를 날릴 수 없을까요?

5seonjae replied:
멘토님께서 말씀해주신대로 `productService.getProductById(productId)` 에서 JPA 가 Product 만 먼저 조회하고, 그 뒤에 `product.getOptions().get(0)` 를 호출하면서 LAZY 컬렉션을 초기화하기 때문에

첫 번째 쿼리: `SELECT * FROM products WHERE id = ?`

두 번째 쿼리: `SELECT * FROM options WHERE product_id = ?`

총 두 번의 SQL이 날아갑니다.

<img width="1753" height="578" alt="image" src="https://github.com/user-attachments/assets/bd75d9d7-3da2-40fd-ac01-a391de4f0b89" />

쿼리를 한 번만 날리기 위해서는 Fetch Join 이나 `@EntityGraph` 를 사용하면 해결됩니다.

두 방법 중 저는 `@EntityGraph` 을 사용했습니다.

```java
public interface ProductRepository extends JpaRepository<Product, Long> {

    @EntityGraph(attributePaths = "options")
    Optional<Product> findById(Long id);
}
```

이렇게 수정을 한다면 `findById` 호출 시 옵션을 JOIN 해 와서, 뒤에 `getOptions()` 해도 추가 쿼리가 없습니다.

<img width="1765" height="535" alt="image" src="https://github.com/user-attachments/assets/7dc56a40-f3b9-45be-8ea1-6786abdc2e2d" />

Gyuchool commented on src/main/java/gift/entity/KakaoTokens.java:
```diff
@@ -0,0 +1,70 @@
+        return new KakaoTokens(
+            kakaoTokenResponseDto.accessToken(),
+            kakaoTokenResponseDto.refreshToken(),
+            LocalDateTime.now().plusSeconds(kakaoTokenResponseDto.expiresIn()),
+            LocalDateTime.now().plusDays(kakaoTokenResponseDto.refreshTokenExpiresIn())
```
정확히 일치시킬 필요까진 없겠지만, 지금처럼하면 now()호출 시점이 달라서 `현재`라는 기준이 다르게 설정되겠네🙄

5seonjae replied:
좋은 지적 감사합니다! 😃 
지금은 `LocalDateTime.now()` 를 두 번 호출하면서 액세스 만료 시간과 리프레시 만료 시간의 기준 시점이 미세하게 달라질 수 있습니다. 
이를 방지하기 위해 한 번만 호출한 값을 기준으로 두 필드를 계산하도록 수정했습니다.

```java
public static KakaoTokens from(KakaoTokenResponseDto dto) {
    LocalDateTime now = LocalDateTime.now();    // 기준 시점을 한 번만 얻음
    return new KakaoTokens(
        dto.accessToken(),
        dto.refreshToken(),
        now.plusSeconds(dto.expiresIn()),        // 같은 기준
        now.plusDays(dto.refreshTokenExpiresIn())// 같은 기준
    );
}
```

이렇게 하면 “현재(now)” 라는 기준이 완전히 일치하므로, 두 필드 간에 미묘한 시간 오차가 사라집니다.

Gyuchool commented on src/main/java/gift/entity/KakaoTokens.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public boolean isAceessExpired() {
+        return accessExpiresAt.isBefore(LocalDateTime.now().plusSeconds(60));
```
60초를 더하는 이유가 무엇일까요?

5seonjae replied:
`isAccessExpired()` 에서 `now().plusSeconds(60)` 처럼 60초 ( 1분 ) 여유 시간을 더해 만료 여부를 판단하는 이유는 다음과 같습니다.

1. 프로액티브 리프레시 ( 사전 갱신 )

- 토큰이 정확히 만료된 순간에 갱신을 시도하면, 네트워크 지연이나 클라이언트·서버 간 시계 ( clock ) 차이로 인해 “만료 직후”의 요청이 실패할 위험이 있습니다.
- 그래서 만료 시점 1분 전부터 “만료 임박”으로 간주하고, 미리 리프레시를 트리거하도록 했습니다.

2. 안정적인 API 호출 보장

- 남은 유효 시간이 짧으면, 그 사이에 전송되는 Kakao API 호출이 중간에 401 에러를 낼 수 있습니다.
- 최소 60초 이상의 유효 기간을 확보하고 요청을 보내도록 미리 갱신하면, 갑작스러운 만료로 인한 실패 케이스를 줄일 수 있습니다.

3. 클럭 스큐 및 재시도 로직 대비

- 분산 환경에서 서버 간 시계가 완벽히 동기화되지 않을 수 있고, 리프레시 도중에도 재시도 로직이 필요할 수 있는데, 남은 시간 버퍼가 있으면 이런 복합 상황에서도 안전합니다.

이렇게 하면 실제 만료 시점 근처의 예외나 무중단 리프레시 실패를 방지할 수 있어 안정성이 한층 높아집니다.
감사합니다! 😊

Gyuchool replied:
버퍼로 60초를 잡아서 미리 갱신하기 위함이군요?! 그렇다면 매직넘버를 제거해도 좋겠네요!
설명을 듣기 전까지 60초를 왜 추가하는지 이해하기 어려울 것 같아요!

Gyuchool commented on src/main/java/gift/entity/Order.java:
```diff
@@ -0,0 +1,91 @@
+    private Long id;
+
+    @Column(nullable = false)
+    @Min(value = 1, message = "수량은 1개 이상이어야 합니다.")
+    @Max(value = 100_000_000, message = "수량은 1억 개 미만이어야 합니다.")
```
엔티티에 validation을 거는 것과 생성자에서 검증하는 것은 어떤 차이가 있을까요?

5seonjae replied:
좋은 질문 감사합니다 😊

엔티티에 어노테이션 기반 검증을 거는 것과, 생성자에서 직접 if문으로 검증하는 방식의 주요 차이는 다음과 같습니다.

### 1. 엔티티 어노테이션 ( `@Min`, `@Max`, ... )

- 스프링 MVC 컨트롤러의 `@Valid`, JPA 의 `@Valid` 트리거, 혹은 수동으로 Validator 를 호출할 때 실행
- 주로 입력값 검증 ( DTO → 엔티티 매핑 전후 ) 을 위해 사용
- 장점
  - 선언적이고 재사용 가능
  - 메시지 코드, 그룹별 검증 기능 활용 가능
- 단점
  - 검증을 트리거하지 않으면 동작하지 않음
  - JPA 저장 시점에만 검증하도록 설정하면, 도메인 생성 시점과 저장 시점이 달라질 수 있음

### 2. 생성자 내부 검증

- 생성자 ( `new Order(...)` ) 가 호출되는 즉시 실행
- **도메인 불변식 ( invariant )** 을 보장하기 위해 사용
- 장점
  - 엔티티 인스턴스가 “절대 잘못된 상태”로 생성되지 않도록 강제
  - 언제나 실행되므로, 검증 누락 위험이 없음
- 단점
  - 메시지 커스터마이징이나 그룹화 등의 부가 기능을 직접 구현해야 함
  - 반복 코드가 많아질 수 있음

### 결론

**사용자 입력**에 대해서는 엔티티 어노테이션을 사용하고, **도메인**에 대해서는 생성자 ( 또는 팩토리 메서드 ) 에서 직접 검사하는 것이 맞는 것 같습니다.

이렇게 입력 검증과 도메인 무결성 보장을 책임 범위에 맞게 분리하면, 유연하면서도 안전한 검증 체계를 갖출 수 있습니다.
이 방식대로 엔티티의 검증 로직을 생성자 내부 검증 방식으로 변경하도록 하겠습니다.
추가 의견 언제든 환영합니다! 🙏

Gyuchool commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -0,0 +1,85 @@
+@Service
+public class KakaoMessageService {
+
+    @Value("${kakao.api-url}")
```
해당 설정이 필요할까요?
생성자 주입과 필드 주입의 차이가 무엇일까요?

5seonjae replied:
리뷰 감사합니다! 지적하신 대로 apiUrl 에 대한 `@Value` 를 필드와 생성자에 중복으로 선언할 필요는 없습니다.
일관되게 생성자 주입만 사용하면 다음과 같은 장점이 있습니다.

1. 불변성 보장

- final 필드에 의존성을 바로 할당할 수 있어, 한 번 주입된 값이 절대 바뀌지 않습니다.

2. 테스트 용이성

- 단위 테스트 시 스프링 컨텍스트 없이도 `new KakaoMessageService(...)` 생성자를 통해 직접 값을 넘겨줄 수 있습니다.

3. 의존 관계 명시

- 클래스가 어떤 외부 값을 필요로 하는지 생성자 시그니처만 봐도 한눈에 파악할 수 있어 가독성이 올라갑니다.

반면에 필드 주입은 리플렉션을 통해 값을 주입하기 때문에 final 선언이 불가능하고 테스트 시에도 반드시 스프링 컨텍스트가 필요해 번거롭습니다.

또한 `templateId` 역시 같은 방식으로 생성자 주입으로 옮겨 두면, 클래스 내부에 주입 방식이 통일되어 유지보수성이 좋아집니다.

```java
@Service
public class KakaoMessageService {

    private final WebClient kakaoClient;
    private final ObjectMapper objectMapper;
    private final Long templateId;

    public KakaoMessageService(
        @Value("${kakao.api-url}") String apiUrl,
        @Value("${kakao.template-id}") Long templateId,
        WebClient.Builder builder,
        ObjectMapper objectMapper
    ) {
        this.kakaoClient  = builder.baseUrl(apiUrl).build();
        this.objectMapper = objectMapper;
        this.templateId   = templateId;
    }

    public void sendOrderMemo(Order order, String accessToken) {
        // ...기존 로직...
    }
}
```

`@Value("${kakao.api-url}")` 와 `@Value("${kakao.template-id}")` 를 모두 생성자 파라미터로 옮겨 필드 주입을 제거하고 apiUrl 필드 위에 붙어 있던 `@Value` 애노테이션과 `final String apiUrl` 선언은 삭제했습니다.

이렇게 하면 의존성 주입 방식이 모두 생성자 주입으로 통일되어, 클래스의 역할과 필요 값을 명확히 드러낼 수 있습니다. 추가 의견 언제든 환영합니다! 😊

Gyuchool commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -0,0 +1,85 @@
+        this.objectMapper = objectMapper;
+    }
+
+    @Value("${kakao.template-id}")
```
여기는 필드 주입을 하는 이유가 있을까요?

5seonjae replied:
https://github.com/next-step/spring-gift-order/pull/182#discussion_r2244440996

위의 답변에 이에 대한 내용을 같이 작성했습니다!

Gyuchool commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -0,0 +1,85 @@
+                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {})
+                .block();
+
+            if (!Objects.equals(resp.get("result_code"), 0)) {
```
0이 의미하는게 무엇일까요? 매직넘버를 제거해볼까요?

5seonjae replied:
지적해 주신 대로, 여기서 0은 카카오톡 메모 전송 API 의 result_code 가 “정상 성공”임을 뜻하는 값입니다. 
매직 넘버를 그대로 쓰면 의미가 불명확하니, 아래처럼 상수로 추출해 가독성을 높였습니다.

```java
@Service
public class KakaoMessageService {
    private static final int KAKAO_SUCCESS_CODE = 0;

    // …생략…

    public void sendOrderMemo(Order order, String accessToken) {
        // …Form 생성, 요청…

        Map<String, Object> resp = kakaoClient.post()
            // …요청 설정…
            .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {})
            .block();

        Integer resultCode = (Integer) resp.get("result_code");
        if (!Objects.equals(resultCode, KAKAO_SUCCESS_CODE)) {
            throw new ExternalServiceException("카카오톡 전송 실패: " + resp);
        }
    }
}
```

이렇게 하면 숫자 하나만 보고도 “성공 코드”임을 바로 이해할 수 있어 가독성이 훨씬 좋아집니다. 감사합니다! 😊

Gyuchool commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,79 @@
+        orderRepository.deleteById(id);
+    }
+
+    private void validateMember(Member member) {
```
member가 null인 경우가 있을까요? resolver에서 검증했을텐데 추가한 이유가 궁금합니다:)

5seonjae replied:
좋은 지적 감사합니다! 😊

맞습니다. `LoginMemberArgumentResolver` 단계에서 이미 Member 를 꺼내면서 인증 실패 시 예외가 터지기 때문에, 컨트롤러나 서비스에 들어오는 member 가 null 이 될 가능성은 사실상 없습니다.

그럼에도 불구하고 서비스 레이어에 `validateMember(member)` 를 한 번 더 둔 이유는 다음과 같습니다.

1. 방어적 프로그래밍 ( Defense in Depth )

- 서비스 메서드는 컨트롤러뿐 아니라, 스케줄러나 다른 내부 API, 테스트 코드 등 여러 경로에서 호출될 수 있습니다.
- 리졸버 없이 직접 서비스 메서드를 호출할 때 member 가 실수로 null 이 넘어오면, NPE 가 아닌 **명확한 InvalidMemberException**으로 잡아주기 위함입니다.

2. 책임 경계 분명히 하기

- 컨트롤러에서 바인딩·인증을 담당하더라도, 서비스는 **“도메인 로직의 진입점”**으로서 입력값 ( 여기서는 member ) 의 유효성을 스스로 검사해야 책임이 명확해집니다.
- 따라서 “서비스가 받는 파라미터는 항상 유효하다”는 전제 하에서 로직을 작성할 수 있어, 유지보수가 더 수월해집니다.

3. 확장 대비

- 앞으로 OrderService 를 외부 API나 배치 처리 등에서 재사용하게 될 때, `@LoginMember` 가 붙은 컨트롤러 없이 바로 `addOrderForMember(null, dto)` 가 호출되는 실수를 방지할 수 있습니다.

요약하자면, 리졸버 단계와는 별개로 서비스 레이어에서도 입력 검증을 한 번 더 둬서 “서비스 메서드 입장에서는 member 가 절대 null 일 수 없다” 라는 불변식을 명확히 하고, 예외가 발생할 때에도 일관된 예외 타입 ( InvalidMemberException ) 으로 처리하기 위함입니다.

혹시 추가 의견 있으시면 편하게 말씀 부탁드립니다! 🙏

Gyuchool commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,79 @@
+        validateMember(member);
+        orderRepository.findById(id)
+            .orElseThrow(() -> new NoSuchElementException("주문을 찾을 수 없습니다."));
+        orderRepository.deleteById(id);
```
deleteById()와 delete()는 어떤 차이가 있을까요?

5seonjae replied:
좋은 질문 감사합니다! 😊

`deleteById(...)` 와 `delete(...)` 의 차이를 정리해 보면 다음과 같습니다.

| 메서드 | 파라미터 | 내부 동작 | 언제 쓰면 좋은가 |
| -- | -- | -- | -- |
| `deleteById(ID id)` | 엔티티의 식별자 | 1. `findById(id)` 호출 → 엔티티 조회<br>2. 조회된 엔티티를 `remove()` | id 만 알고 있을 때, 편리하게 삭제 가능 |
| `delete(T entity)` | 엔티티 인스턴스 | 1. 전달된 엔티티가 영속 상태라면 바로 `remove()`<br>2. ( 준영속 상태라면 `merge()` 후 `remove()` ) | 이미 로드된 ( 또는 관리 중인 ) 엔티티를 삭제할 때 |

- 조회 쿼리 차이

  - `deleteById` 는 내부에서 `findById` 를 한 번 더 실행하기 때문에, 트랜잭션 내에 해당 엔티티가 영속성 컨텍스트에 없으면 추가 SELECT 가 발생합니다.
  - `delete(entity)` 는 이미 영속 상태인 인스턴스를 넘기면 바로 DELETE 만 호출하므로, 불필요한 조회를 피할 수 있습니다.

- 예외 처리 차이

  - `deleteById(id)` 는 id 에 해당하는 엔티티가 없으면 `EmptyResultDataAccessException` 을 던집니다.
  - `delete(entity)` 는 **entity 가 null** 이거나 식별자가 비어 있으면 `IllegalArgumentException` 이 발생할 수 있습니다.

### 결론

지금처럼 “존재 여부를 먼저 findById 로 검증 → deleteById” 패턴은 400/404 취지에 맞춰 “없으면 바로 예외”를 던지고 싶을 때 적합합니다.

그래서 쿼리 최적화를 위해 `delete` 로 수정하려고 합니다.

```java
public void deleteOrderForMember(Member member, Long id) {
        validateMember(member);
        Order order = orderRepository.findById(id)
            .orElseThrow(() -> new NoSuchElementException("주문을 찾을 수 없습니다."));
        orderRepository.delete(order);
    }
```

이렇게 두 메서드의 내부 동작과 사용 시기를 구분하면, 불필요한 쿼리를 줄이거나 예외 타입을 의도대로 선택하기가 훨씬 쉬워집니다.
감사합니다! 🙏

Gyuchool commented on src/test/java/gift/OrderServiceTest.java:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    @ParameterizedTest(name = "[page={0}] quantity={1}, message={2}")
+    @CsvSource({
```
csvSource 사용👍
다만, 매개변수 명으로 충분히 표현되니 주석은 제거해도 좋겠네요:)

Gyuchool commented on src/test/java/gift/OrderServiceTest.java:
```diff
@@ -0,0 +1,227 @@
+import org.springframework.data.domain.Sort;
+
+@ExtendWith(MockitoExtension.class)
+public class OrderServiceTest {
```
테스트 꼼꼼히 잘 작성하셨네요👍

Gyuchool reviewed (approved):
안녕하세요 선재님!
빠르게 피드백 반영 및 의견을 주셨군요👍👍 
기능과 테스트 잘 동작해서 여기서 머지하도록 하겠습니다:) 다음 단계에서 뵐게요!

몇 가지 코멘트를 남겨두었으니 다음 단계 진행하시면서 반영 및 의견 주시면 좋을 것 같습니다.🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
다음 미션도 화이팅입니다!! 💪
